### PR TITLE
Minor addition to VectorFEMassIntegrator

### PR DESCRIPTION
**Pull request by stefanozampini**
Add support for tensor permeability for mixed forms

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/1068 | @stefanozampini | @v-dobrev | @homijan + @delyank + @mlstowell | 12/13/19 | 01/16/20 | ⌛due 01/23/20 |

### CODE REVIEW

**Comment by stefanozampini** [comment]
@tzanio @v-dobrev any reason why this is not yet merged to master?

**Comment by stefanozampini** [comment]
@tzanio @v-dobrev It would be nice to have this in MFEM 4.1

**Comment by stefanozampini** [comment]
@tzanio @v-dobrev A gentle reminder to include this already approved PR in MFEM 4.1

**Comment by delyank** [comment]
This PR adds new code and slightly modifies existing code. `VectorFEMassIntegrator::AssembleElementMatrix2` is modified (in the case of vector spaces for both trial and test functions) in a way that matches almost line by line (with minor adjustments) the respective portion of `VectorFEMassIntegrator::AssembleElementMatrix`. It now allows vector and matrix coefficients in that case and, as a side effect, the already existing code for scalar coefficients is made nicer.

I don't have a test case with different trial and test spaces. It seems that if `VectorFEMassIntegrator::AssembleElementMatrix2` is invoked with test=trial it will operation by operation recreate `VectorFEMassIntegrator::AssembleElementMatrix`. So it seems safe and on this stage I **approve** of the PR, but have in mind I haven't executed the autotest yet.

**Comment by tzanio** [comment]
@delyank, do you mind clicking on the approve button (if you do still approve :)) ?

File: fem/bilininteg.cpp [hunk 1]
```diff
@@ -1688,71 +1688,15 @@ void VectorFEMassIntegrator::AssembleElementMatrix(
...
+   AssembleElementMatrix2(el, el, Trans, elmat);
 }
 
 void VectorFEMassIntegrator::AssembleElementMatrix2(
    const FiniteElement &trial_fe, const FiniteElement &test_fe,
    ElementTransformation &Trans, DenseMatrix &elmat)
 {
-   if ( test_fe.GetRangeType() == FiniteElement::SCALAR && VQ )
+   if ( ( test_fe.GetRangeType() == FiniteElement::SCALAR && 
+          trial_fe.GetRangeType() == FiniteElement::VECTOR ) && VQ )
```
Comment (homijan) [inline fem/bilininteg.cpp]: The option using VQ with FiniteElement::SCALAR should be extended by MQ and squezed into one block starting at line 1757.

File: fem/bilininteg.cpp [hunk 2]
```diff
@@ -1906,23 +1855,35 @@ void VectorFEMassIntegrator::AssembleElementMatrix2(
...
             }
+            AddMult_a_ABt(w,test_vshape,trial_vshape,elmat);
          }
       }
    }
+   else
+   {
+      mfem_error("VectorFEMassIntegrator::AssembleElementMatrix2(...)\n"
+                 "   is not implemented for this mix of\n"
+                 "   FiniteElement::SCALAR and FiniteElement::VECTOR bases.");
```
Comment (homijan) [inline fem/bilininteg.cpp]: It would be useful to add the FiniteElement::SCALAR range for both trial and test bases.

File: fem/bilininteg.cpp [hunk 3]
```diff
@@ -1688,93 +1688,36 @@ void VectorFEMassIntegrator::AssembleElementMatrix(
...
+      DenseMatrix K(MQ ? MQ->GetVDim() : 0, MQ ? MQ->GetVDim() : 0);
 #else
       trial_vshape.SetSize(trial_dof, dim);
       shape.SetSize(test_dof);
-      D.SetSize(dim);
+      D.SetSize(VQ ? VQ->GetVDim() : 0);
+      K.SetSize(MQ ? MQ->GetVDim() : 0, MQ ? MQ->GetVDim() : 0);
 #endif
 
-      elmat.SetSize (test_dof, trial_dof);
```
Comment (homijan) [inline fem/bilininteg.cpp]: H1 is used to discretize each of the vector components, and so, elmat has to have size (dim*test_dof, trial_dof) in the case of SCALAR-VECTOR mix of basis.

File: fem/bilininteg.cpp [hunk 4]
```diff
@@ -1794,42 +1737,103 @@ void VectorFEMassIntegrator::AssembleElementMatrix2(
...
+            w *= Q -> Eval (Trans, ip);
+         }
 
          for (int d = 0; d < dim; d++)
          {
             for (int j = 0; j < test_dof; j++)
             {
                for (int k = 0; k < trial_dof; k++)
                {
-                  elmat(j, k) += D[d] * shape(j) * trial_vshape(k, d);
```
Comment (homijan) [inline fem/bilininteg.cpp]: This is missing a shift in the component of the vector function.

**Comment by homijan** [comment]
I made the change for testing of the AssembleElementMatrix2 version. I have reverted back the once-sampled version. It works for VECTOR and SCALAR range type.

File: fem/bilininteg.cpp (hunk 3 above)
Comment (homijan) [reply fem/bilininteg.cpp]: This was actually an old bug.

File: fem/bilininteg.cpp (hunk 4 above)
Comment (homijan) [reply fem/bilininteg.cpp]: This was an old bug, now it has been corrected.

**Comment by tzanio** [comment]
Merged in `next` for testing ...

**Comment by v-dobrev** [comment]
Added @mlstowell as a reviewer to take a look at the changes by @homijan.

@stefanozampini, @delyank, can you please also take a look at the latest state and let me know if it looks good to you?

File: fem/bilininteg.hpp [hunk 5]
```diff
@@ -2133,6 +2136,7 @@ class VectorFEMassIntegrator: public BilinearFormIntegrator
    Vector shape;
    Vector D;
    DenseMatrix K;
+   DenseMatrix partelmat;
```
Comment (mlstowell) [inline fem/bilininteg.hpp]: I don't see this used in the source file.  Is it still needed?

File: fem/bilininteg.cpp (hunk 3 above)
Comment (mlstowell) [reply fem/bilininteg.cpp]: Actually, I believe this was not a bug but an undocumented feature.  The code actually assumed `vdim=1` in the scalar field and computed the bilinear form `(v.w, u)` where `v` is the vector coefficient, `w` is the Nedelec or RT basis function, and `u` is the scalar basis function.  It is confusing to have this operator here and I could not find any examples or mini apps that use it so it is probably best to remove this (as this PR does).  Incidentally, this confusing functionality is now available, and documented, in `MixedDotProductIntegrator` if anyone needs it.

File: fem/bilininteg.cpp [hunk 6]
```diff
@@ -1758,29 +1758,29 @@ void VectorFEMassIntegrator::AssembleElementMatrix2(
...
 #else
       trial_vshape.SetSize(trial_dof, dim);
       shape.SetSize(test_dof);
-      D.SetSize(dim);
+      D.SetSize(VQ ? VQ->GetVDim() : 0);
+      K.SetSize(MQ ? MQ->GetVDim() : 0, MQ ? MQ->GetVDim() : 0);
 #endif
 
-      elmat.SetSize (test_dof, trial_dof);
+      elmat.SetSize (dim*test_dof, trial_dof);
```
Comment (mlstowell) [inline fem/bilininteg.cpp]: This line assumes that `vdim==dim` but we should verify this with an `MFEM_ASSERT` or something similar.

Comment (mlstowell) [reply fem/bilininteg.cpp]: @v-dobrev just reminded me we don't have access to `vdim` inside the integrators so please disregard this comment.

**Comment by delyank** [comment]
> Added @mlstowell as a reviewer to take a look at the changes by @homijan.
> 
> @stefanozampini, @delyank, can you please also take a look at the latest state and let me know if it looks good to you?

I'm fine with the changes, if this (at least by exception) is an acceptable way to modify a pull request. The overall modifications are considerable. I hope Stefano is OK with this as he is the author of the pull request and the modified version still serves the purpose that he desires to achieve with this pull request.

**Comment by v-dobrev** [comment]
Thanks @delyank! Yes, this way of adding more changes to a PR is not the right way -- let's make a one-time exception here, if @stefanozampini is also OK.

**Comment by stefanozampini** [comment]
I'm ok with the changes. Thanks

**Comment by homijan** [comment]
Thanks. Will be better next time.

**Pull request merged**